File: apps/cli/src/caspian_cli/intent.py
from dataclasses import dataclass, field
from typing import Literal, Union


@dataclass(frozen=True)
class ChannelsAdd:
    channel: str
    via: Literal["hosted", "self-host"]
    display_name: str = ""
    bot_token: str = ""
    webhook_url: str = ""
    inbound: bool = True


@dataclass(frozen=True)
class ChannelsLs:
    pass


@dataclass(frozen=True)
class Call:
    """The only mutate/send intent. `id` is a catalog id (`post`, `telegram.send-photo`)."""

    id: str
    args: dict = field(default_factory=dict)


@dataclass(frozen=True)
class CatalogList:
    pass


@dataclass(frozen=True)
class CatalogSearch:
    query: str


@dataclass(frozen=True)
class CatalogGet:
    id: str


@dataclass(frozen=True)
class ThreadsLs:
    channel: str = ""


@dataclass(frozen=True)
class ThreadsTail:
    thread_id: str = ""


Intent = Union[
    ChannelsAdd,
    ChannelsLs,
    Call,
    CatalogList,
    CatalogSearch,
    CatalogGet,
    ThreadsLs,
    ThreadsTail,
]

File: packages/cli/src/intent.ts
/**
 * CLI intents — programs as data. Argv desugars here; plan.ts denotes.
 *
 * Call fields are closed (no string bag). Absent flags are empty strings so
 * the type is total; planIntent rejects combinations the catalog cannot run.
 */
export type Via = "hosted" | "self-host"

export type ChannelsAdd = {
  readonly _tag: "ChannelsAdd"
  readonly channel: string
  readonly via: Via
  readonly display_name: string
  readonly bot_token: string
  readonly webhook_url: string
  readonly inbound: boolean
}

export type ChannelsLs = {
  readonly _tag: "ChannelsLs"
}

export type Call = {
  readonly _tag: "Call"
  readonly id: string
  readonly thread_id: string
  readonly text: string
  readonly file: string
}

export type CatalogList = {
  readonly _tag: "CatalogList"
}

export type CatalogSearch = {
  readonly _tag: "CatalogSearch"
  readonly query: string
}

export type CatalogGet = {
  readonly _tag: "CatalogGet"
  readonly id: string
}

export type ThreadsLs = {
  readonly _tag: "ThreadsLs"
  readonly channel: string
}

export type ThreadsTail = {
  readonly _tag: "ThreadsTail"
  readonly thread_id: string
}

export type Login = {
  readonly _tag: "Login"
  readonly open: boolean
  readonly gateway: string
}

export type InitKind = "cli" | "project" | "agent"

/** Bare `caspian init` asks; a kind skips the question. */
export type InitTarget = InitKind | "ask"

export type Init = {
  readonly _tag: "Init"
  readonly kind: InitTarget
  readonly open: boolean
  readonly gateway: string
  readonly force: boolean
  /** For project: folder to write ./.env into. Empty means ask (default: cwd). */
  readonly path: string
  /** For project: scaffold a new TS/Python SDK app — not implemented yet. */
  readonly fresh: boolean
}

export type Intent =
  | ChannelsAdd
  | ChannelsLs
  | Call
  | CatalogList
  | CatalogSearch
  | CatalogGet
  | ThreadsLs
  | ThreadsTail
  | Login
  | Init
